### DISCUSSION TITLE
我使用了旁側的方式識別REALITY，誤報率是4.7%

### DISCUSSION THREAD

**Discussion by nursery01** [XTLS/Xray-core#2161] [General]
代碼在這裏[https://github.com/nursery01/GFW-system/tree/main/TLS/passive%20detection](https://github.com/nursery01/GFW-system/tree/main/TLS/passive%20detection)

防止被這種旁側識別的解決方案就是

1.相近的ip(其實就是R神説過的話)

2.偷自己的網站(100%無法通過這種旁側方法識別，因爲就是自己的)

REALITY自身應該是沒有什麽安全問題能被識別的，~~其實是我也不懂得如何識別~~

**Comment by RPRX** [comment]
大概看了下，我比较好奇的是，你捕获到的流量是静态的，Geo 数据库也是静态的，为什么对于同一份流量，误报率越来越高？

此外，稍大点的网站都有很多 IP，现在还流行 CDN、边缘计算之类的，离你更近，IP 更杂。你可以多换几个 DNS，清空下缓存，测一测。

想针对性识别“偷别人”，我有一些其它方法，~~不过看起来伊朗 GFW 还没研究出来 / 还没部署，我就先不说了。~~

**Comment by nursery01** [comment]
> 大概看了下，我比较好奇的是，你捕获到的流量是静态的，Geo 数据库也是静态的，为什么对于同一份流量，误报率越来越高？

它是對伺服器的IP進行ASN查詢得到第一個ASN，然後對sni進行dns查詢得到IP再進行ASN查詢得到第二個ASN，然後對比ASN。我第四次和第五次查詢均在5%左右，並沒有越來越高。所以不是bug，產生波動的原因是dns解析出不同的IP

實際上可能誤報率更高，這只是小規模實驗環境，即使誤報率固定5%，全球百萬千萬個網站也會誤封很多

**Reply by RPRX** [reply]
那我猜对了，如果涉及到实时查询 DNS，你忽略了一个重要的问题：**这种检测程序不能在本机跑。**

因为首先你本机的 DNS 没变，其次你正常上网时，DNS 在本机有缓存，紧接着你在本机跑检测程序，大概率就会解析出相近的 IP。

所以你的测试方法是不对的，导致数据失去了参考价值。**你应该用另一台电脑，设另一个 DNS，在另一个网络下跑检测程序。**

**Reply by RPRX** [reply]
或者同一台电脑，把 DNS 缓存清空，~~如果清得干净的话~~，设另一个 DNS，在手机网络下跑检测程序（最好是另一个运营商）。

总之涉及到实时查询 DNS 就比较复杂了，无法完全复现 GFW 所面临的情况，**比如它应该查哪个 DNS？通过哪个网络去查？**

如果用户明文查 DNS，GFW 会知道他向谁查、查询结果，但对密文查 dns、本地缓存、改 hosts 或应用指定 IP 等就不好使。

**Reply by RPRX** [reply]
我向可能不了解的人解释一下：由于有多 IP、CDN、边缘计算等，你查不同的 DNS、用不同的网络，返回的结果都可能不同。

**Reply by nursery01** [reply]
我這個在本機跑的檢測程式其實就是在模擬GFW。假如說我人在海南擷取資料得到ASN，然後這時候GFW在新疆進行DNS查詢得到不同的ASN，然後GFW說我是代理，肯定冤枉啊。

但是你的意思是說我在海南A區解析ASN，然後GFW要在海南B區解析ASN，這樣不會產生DNS快取，影響檢測結果，這個也是對的

我覺得真的這樣測試的話，嗯🤔......我這邊看來主要是套了cdn的網站ip變動很大，如果把cdn ip列為白名單直接放過的話誤報率會極大幅下降，因為cdn ip不可能是reality，分散式系統的IP也變動大

再去測我現在也懶得測了，至少中國我不太認為會用這種方式部署，至於伊朗(我才不管，我先封了再說)

**Reply by RPRX** [reply]
当然不是海南到新疆那么远，而是不同的 DNS 本来就有不同的解析倾向，包括对不同的运营商网络返回不同的结果。所以在同一个电脑上什么都不改，甚至还有 DNS 缓存时直接跑检测，误报率肯定比不在你电脑上的 GFW 跑的检测小得多，这个数据是严重偏离实际情况的。IP 白名单的话，网站、CDN、边缘计算等经常加个节点、撤个节点，所以也不太好使。

重测最大的意义在于：**防止什么都不懂的小白拿这个数据当参考，甚至去宣传。**
XTLS 刚出时有个人用非 TLS 流量去测性能，结果是和 TLS 差不多，我想：这不一眼看出来不是 TLS 流量吗，这谁会信？
结果是有不少小白信了，还经常在各个群里发，所以我现在非常注意这种情况，必须及时纠正。

就是说，你既然把这样的数据放出来了，就应当对数据的准确性负责，否则会被人引用，给我们造成不少不必要的麻烦。

**Reply by nursery01** [reply]
> 就是说，你既然把这样的数据放出来了，就应当对数据的准确性负责，否则会被人引用，给我们造成不少不必要的麻烦。

啊，原來你的擔憂是這樣，但是我在這個專案創建初期就在readme裡面寫「不保證科學性」，短期小實驗的東西肯定會和全局有很大的偏差

又或許是我當了叛徒做了這個不好的東西傷了大神的心，但是reality本身沒有問題，不太可能被識別，說簡單點就是很安全，至於旁側、社會學、還是心情不好直接掐斷是另外的層面了，代理作者也只能改改ISO第7層，控制底層的GFW實在不允許也不是第7層的問題

這個玩具我以後也不想更新它了，因為也不好玩，我去整點其它的

**Reply by RPRX** [reply]
> 啊，原來你的擔憂是這樣，但是我在這個專案創建初期就在readme裡面寫「不保證科學性」，短期小實驗的東西肯定會和全局有很大的偏差

这个之前我就看到了（但是不进主页就看不到，所以会有人看不到），然而关心这些事的人水平参差不齐，有很多是完全的小白，即使你这么写，他们也会拿来当参考、宣传材料，你经历多了就懂了。

> 又或許是我當了叛徒做了這個不好的東西傷了大神的心

这倒没有，围绕这方面的研究、实验，提供参考也是有价值的。只是测试方法要科学、数据要真实，这样才真正有参考价值。

> 這個玩具我以後也不想更新它了，因為也不好玩，我去整點其它的

再测一次应该不麻烦吧？因为我也比较好奇用我说的方法测，测出来的误报率是多少，大家应该都想参考一下。

**Reply by nursery01** [reply]
我今天再對昨天同一份舊資料用同一個dns供應商進行測試是10.7%誤報率，然後更換dns并且刷新dns快取后是9.5%誤報率，更換dns供應商沒發現會導致誤報率大幅波動，倒是時間過久了會導致誤報率上漲

可能是服務ip換了，而且昨天和今天剛好是月底31日和月初1日

誤報的網域大多數都是cdn或者api

**Reply by RPRX** [reply]
这就给大家提供了参考价值：偷别人可以选这类域名，GFW 若想通过 DNS 记录来封，没有那么简单，只能维护很多动态 IP 白名单。

我觉得和月底月初的关系不大，dns 的话你可以换个和之前区别较大的，比如换成某个加密的 dns，**此外需要换个运营商再测试。**

至于海南 A 区、海南 B 区，不同的地理位置、源 IP 当然也会有影响，~~但这个需要你带着电脑跑~~，有空你可以去星巴克试试。
